ArC: support interception of producers and synthetic beans

Ladicek opened the pull request:
Fixes #3532

Ladicek commented:
Submitting this as a draft because:

- there are 3 commits that should eventually be squashed into 1
- there are 2 major TODOs in the code that should be solved in one way or another
- documentation is missing

At the same time, I believe this has reached enough maturity and can be reviewed.

manovotn commented:
> After discussing with @mkouba and @manovotn, I made these significant changes:
> 
>     * interception of producer methods and synthetic beans is proxy-based now
> 
>     * replaced `InterceptedObject` and `@SupportsInterception` with `InterceptionProxy` and `@BindingsSource`
> 
> 
> It turns out that a producer method (as well as a synthetic bean) can now inject multiple different `InterceptionProxy` instances (although there's one TODO about that in the code).

Thanks for working on this!
I am planning to give it a deeper look and provide feedback later today or tomorrow.

mkouba commented on independent-projects/arc/runtime/src/main/java/io/quarkus/arc/BindingsSource.java:
```diff
@@ -0,0 +1,36 @@
+ * a method with an identical name, return type and parameter types declared
+ * on the bindings source class.
+ * <p>
+ * The annotations that can be meaningfully present on the {@code value} class are:
```
_"The annotations that can be meaningfully present..."_ This sentence sounds a bit stilted.

Ladicek replied:
Let me rewrite :-)

Ladicek replied:
Done.

mkouba commented on independent-projects/arc/runtime/src/main/java/io/quarkus/arc/BindingsSource.java:
```diff
@@ -0,0 +1,36 @@
+    /**
+     * The class from which interceptor binding annotations should be read.
+     */
+    Class<?> value();
```
I wonder if it would make sense to support multiple binding sources? I.e. either make this annotation repeatable or use  `Class<?>[] value()`...

Ladicek replied:
I'm have no idea how combining multiple bindings sources would work. I would stay away from that if possible.

mkouba replied:
I was thinking about simple merge but you're right that for interceptor bindings with members we would need priorities and such...

mkouba replied:
You can do that ofc but my original idea was about sharing the bindings, e.g. when you declare multiple bindings sources for a class hierarchy.

manovotn replied:
I understand, I just think that might be an overkill, at least initially, as it adds complexity to the implementation with little usage/upside.

If we want to share them, then we could instead allow the single class declared here to have subclasses that we also inspect. That could follow rules very similar to what interceptors already do and you wouldn't run into conflicts with priorities, annotation member values etc. That being said, I still vote for the current design :)

mkouba commented on independent-projects/arc/runtime/src/main/java/io/quarkus/arc/BindingsSource.java:
```diff
@@ -0,0 +1,36 @@
+ * <p>
+ * Interceptor bindings for method named {@code M} with return type {@code R}
+ * and parameter types {@code P1, ..., Pn} on the target class are obtained from
+ * a method with an identical name, return type and parameter types declared
```
How about generics? Imagine the target class `Foo` has a method `<T> void foo(List<T> list);`. I suppose that a method `void foo(List<T> list)` declared on `FooBindings1<T>` is matched but `<B> void foo(List<B> list)` declared on `FooBindings2`?

Ladicek replied:
Good point. I'm actually not sure how that would work today, I need to write a test.

Ladicek replied:
Changed the `BindingsDiscovery` class to not use `ClassInfo.method()`. Instead, we iterate through all methods of the class, comparing the method name, return type, parameter types, and the `static` flag. Also, discovering bindings on superclasses of the bindings source is now gone.

mkouba replied:
> comparing ....return type, parameter types

By means of `Object#equals(Object)`?

Ladicek replied:
Yes, it's `Type.equals()`.

mkouba commented on independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanConfiguratorBase.java:
```diff
@@ -278,6 +285,73 @@ public THIS startup() {
+     * @param targetClass the target class, must not be {@code null}
+     * @return self
+     */
+    public THIS supportsInterception(Class<?> targetClass) {
```
Hm, shouldn't we just use the `BeanConfiguratorBase.implClazz`? I.e. `supportsInterception(DotName... bindingsSourceClasses)`... should be enough.

Ladicek replied:
Hmm, yeah, I think this method should be called `injectInterceptionProxy()` or something like that.

It is not super useful I guess, but it allows you to inject 2 interception proxies of 2 different classes and then combine them into a 3rd class, which is the `implClazz` of the synthetic bean. There's a test for this for producers.

Ladicek replied:
It's https://github.com/quarkusio/quarkus/pull/41258/files#diff-f0c3315a9b99e04585136501ea4bc529c710bb6f07f1c6729b3ef0888879ccab

Ladicek replied:
Renamed to `injectInterceptionProxy()`. Also removed the variants accepting a `DotName` target class; instead, they accept `org.jboss.jandex.Type`.

Ladicek replied:
Also removed the `Type` parameter; we now always use the `implClazz` (actually the `providerType`, which is almost always derived from the `implClazz`, but doesn't necessarily have to be).

mkouba reviewed (commented):
I added some comments to the API. Didn't look at the implementation yet...

mkouba commented on independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/interceptors/producer/ProducerWithMultipleInterceptionProxyParamsTest.java:
```diff
@@ -0,0 +1,89 @@
+    @Dependent
+    static class MyProducer {
+        @Produces
+        String produce(InterceptionProxy<MyNonbean1> proxy1, InterceptionProxy<MyNonbean2> proxy2) {
```
This is IMO superweird and should not be allowed. The `InterceptionProxy` should only be used to create a contextual instance of a specific producer method.

Ladicek replied:
I agree it's weird and probably not very useful, but the `InterceptionFactory` equivalent works in Weld just fine:

```java
@Produces
String produce(InterceptionFactory<MyNonbean1> nonbean1,
        InterceptionFactory<MyNonbean2> nonbean2) {
    return nonbean1.createInterceptedInstance(new MyNonbean1()).hello()
        + " | " + nonbean2.createInterceptedInstance(new MyNonbean2()).hello();
}
```

mkouba replied:
That was probably an oversight. I do not believe the `InterceptionFactory` was intended to work this way.

@manovotn WDYT?

Ladicek replied:
OK, let's constrain this down. The type argument to `InterceptionProxy` has to be ... one of the bean types of the bean? Sounds reasonable?

Also, at most one injection point of type `InterceptionProxy<...>` for each producer / synth bean.

Ladicek replied:
Done.

Ladicek committed:
ArC: fix Types.getParameterizedType() in case of a nested generic class

The previous code always used `null` as the owner type, which is incorrect
in case of nested generic classes. This commit fixes that in a not entirely
precise but good enough way.

Ladicek commented:
Also added one extra commit that fixes `Types.getParameterizedType()` in case of a nested generic class.

Ladicek commented:
Also added documentation in `cdi-reference`.

manovotn commented on independent-projects/arc/runtime/src/main/java/io/quarkus/arc/InterceptionProxy.java:
```diff
@@ -0,0 +1,29 @@
+ * interceptor classes; other kinds of interception, as well as {@code @AroundInvoke}
+ * interceptors declared on target classes, are not supported.
+ */
+public interface InterceptionProxy<T> {
```
Maybe we should do this similarly to how CDI defines `InterceptionFactory` - that it is a *bean* that is provided to you by the container and that you can only use it as param of producer method or in synth beans.
We can then validate and test that it doesn't get injected anywhere else (c-tors, fields, ...) which is a user mistake anyway.

manovotn replied:
I am also thinking that the current approach cannot be used with CDI build compat. extension in any way.
If we managed to turn this into a bean that also carries the bindings source declaration with it (pass in `@BindingsSource` as a fake qualifier perhaps?), we might be able to do that as well, right?

Ladicek replied:
> it is a bean that is provided to you by the container and that you can only use it as param of producer method or in synth beans

Agree. I'll rewrite the javadoc, which should be enough, because internally, `InterceptionProxy` already kinda sorta is a bean (a built-in one).

> We can then validate and test that it doesn't get injected anywhere else (c-tors, fields, ...)

I think we already do, see `BuiltinBean.validateInterceptionProxy()`.

> the current approach cannot be used with CDI build compat. extension in any way

Correct.

> If we managed to turn this into a bean that also carries the bindings source declaration with it (pass in `@BindingsSource` as a fake qualifier perhaps?), we might be able to do that as well, right?

Not sure, to be honest. `SyntheticBeanBuilder` has no methods to add injection points or anything like that. And we don't support "injecting" custom parameters into BCEs (yet).

manovotn replied:
> I think we already do, see BuiltinBean.validateInterceptionProxy().

Will do, I probably missed it there.

> Not sure, to be honest. SyntheticBeanBuilder has no methods to add injection points or anything like that. And we don't support "injecting" custom parameters into BCEs (yet).

Hm, right.
We'd probably need to ask users to subclass `SyntheticBeanCreator` or make it extend some new interface so that we knew what the interception proxy requirements are and only then could they use the `Instance` to resolve it. But that sounds a bit wonky, WDYT?

Ladicek replied:
Yeah, agree. I would stay away from BCE here.

Ladicek replied:
Done, I think.

manovotn commented on independent-projects/arc/runtime/src/main/java/io/quarkus/arc/BindingsSource.java:
```diff
@@ -0,0 +1,38 @@
+ * <p>
+ * Method-level interceptor bindings for a method on the target class are equal to
+ * method-level bindings of a method with the same name, return type, parameter
+ * types and {@code static} flag declared on the bindings source class.
```
I see I missed the previous discussion (https://github.com/quarkusio/quarkus/pull/41258#discussion_r1649031766). Still, I am not sure I understand why special case `static` - method name, params and ret. type should suffice?

Ladicek replied:
I was wondering about this and wasn't really sure if modifiers should apply. [Method Modifiers](https://docs.oracle.com/javase/specs/jls/se17/html/jls-8.html#jls-8.4.3) are:

- `public`
- `protected`
- `private`
- `abstract`
- `static`
- `final`
- `synchronized`
- `native`
- `strictfp`

Out of these, I only found `static` makes a meaningful difference. While it is true that you cannot declare the a pair of non-`static` and `static` methods with the same name and signatures, it doesn't feel right to have an interceptor binding on a `static` method on the bindings source class applying to a non-`static` method on the original class.

manovotn replied:
Given that we allow abstract class as a template for bindings, we already have methods with different modifier.
I see what you mean with `static` and it does feel weird but, to me, no less weird than adding a `synchronized` version :shrug: 

If we don't want to allow this, we could add a check and throw an error if we find signature-matching method which differs in these modifiers? Though it is artificial limitation and has IMO no real justification, so I'd just ignore them.

mkouba commented on docs/src/main/asciidoc/cdi-reference.adoc:
```diff
@@ -1069,6 +1069,189 @@ public class NoopAsyncObserverExceptionHandler implements AsyncObserverException
+Also, the intercepted class or any of its intercepted methods should not be `final`.
+If it is, a bytecode transformation will <<unproxyable_classes_transformation,remove>> it, unless disabled.
+
+Finally, the intercepted class should have a zero-parameter constructor.
```
Hm,  but we also try to add a no-args constructor if missing, or?

manovotn replied:
Yes, the sentence below says just that. My comment was just to point to the cause of that restriction (and others mentioned in this section)

Ladicek replied:
Will do.

Ladicek replied:
Done.

mkouba commented on docs/src/main/asciidoc/cdi-reference.adoc:
```diff
@@ -1069,6 +1069,189 @@ public class NoopAsyncObserverExceptionHandler implements AsyncObserverException
+
+The concept of _bindings source_ is a build-time friendly equivalent of `InterceptionFactory.configure()`.
+
+It is common that the type argument of `InterceptionProxy` is the same as the return type of the producer method, but it doesn't have to be.
```
> Anything you produce with InterceptionProxy has to have a type that will fit the return type of the method - if we enforce same type, we could validate it as well, which might be more beneficial?

:+1:

Ladicek replied:
Will remove these 2 lines.

Ladicek replied:
Done.

Ladicek commented on docs/src/main/asciidoc/cdi-reference.adoc:
```diff
@@ -1069,6 +1069,189 @@ public class NoopAsyncObserverExceptionHandler implements AsyncObserverException
+=== Intercepting Producer Methods and Synthetic Beans
+
+By default, interception is only supported for managed beans (also known as class-based beans).
+To support interception of producer methods and synthetic beans, the CDI specification includes an `InterceptionFactory`, which is not supported in Quarkus.
```
Good point, will do.

Ladicek replied:
Done.

Ladicek commented on docs/src/main/asciidoc/cdi-reference.adoc:
```diff
@@ -1069,6 +1069,189 @@ public class NoopAsyncObserverExceptionHandler implements AsyncObserverException
+Finally, the intercepted class should have a zero-parameter constructor.
+If it doesn't, a bytecode transformation will add it, but that isn't always possible.
+
+It is important to notice that in the example above, interceptor bindings are read from the `MyClass` class.
```
Will do.

Ladicek replied:
Done.

Ladicek commented on independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/interceptors/producer/ProducerWithInterfaceInterceptionTest.java:
```diff
@@ -0,0 +1,104 @@
+    @Dependent
+    static class MyProducer {
+        @Produces
+        MyNonbean produce(InterceptionProxy<MyNonbean> proxy) {
```
You're right. Interesting. I'll investigate why that happens, even though that situation should probably not be legal.

manovotn replied:
Right, that makes sense.
Maybe we could add some javadoc to `InterceptionProxy` saying that unless you specify bindings explicitly, the `T` argument is what we scan for bindings?

Ladicek replied:
It's not the type argument, it's the provider type of the bean, but those are the same thing actually :-) But understood, and done.

Ladicek commented on docs/src/main/asciidoc/cdi-reference.adoc:
```diff
@@ -1069,6 +1069,189 @@ public class NoopAsyncObserverExceptionHandler implements AsyncObserverException
+[source,java]
+----
+public MyClass create(SyntheticCreationalContext<MyClass> context) {
+    InterceptionProxy<MyClass> proxy = context.getInjectedReference(new TypeLiteral<>() {}); // <1>
```
Good point. Will do.

Ladicek replied:
Done.

Ladicek commented on docs/src/main/asciidoc/cdi-reference.adoc:
```diff
@@ -1069,6 +1069,189 @@ public class NoopAsyncObserverExceptionHandler implements AsyncObserverException
+The concept of _bindings source_ is a build-time friendly equivalent of `InterceptionFactory.configure()`.
+
+It is common that the type argument of `InterceptionProxy` is the same as the return type of the producer method, but it doesn't have to be.
+A producer method can even declare multiple `InterceptionProxy` parameters and combine them to produce the eventual outcome.
```
Removed.

Ladicek committed:
ArC: make some SubclassGenerator methods easier to reuse

Ladicek committed:
ArC: support interception of producer methods and synthetic beans

Ladicek commented:
I have updated `SubclassGenerator` to remove a substantial part of duplicate code. The `InterceptionProxyGenerator` still has 200 lines for generating the interception subclass, which is a lot, but I don't think I can remove the duplicity without making a big mess. Ideas welcome.

Also marked this PR as ready for review.

geoand commented:
This needs a rebase

Ladicek commented:
Rebased and resolved the docs conflict.

geoand commented:
🙏🏼

Ladicek commented:
@mkouba @manovotn could you please take another look? Thanks!

mkouba commented on docs/src/main/asciidoc/cdi-reference.adoc:
```diff
@@ -1098,6 +1098,186 @@ public class MyService {
+
+In this example, interceptor bindings are read from the `MyClass` class.
+
+Note that `InterceptionProxy` only supports `@AroundInvoke` interceptors declared on interceptor classes.
```
This would deserve an AsciiDoc [admonition](https://docs.asciidoctor.org/asciidoc/latest/blocks/admonitions/), NOTE or IMPORTANT...

Ladicek replied:
Done.

mkouba commented on docs/src/main/asciidoc/cdi-reference.adoc:
```diff
@@ -1098,6 +1098,186 @@ public class MyService {
+By default, interception is only supported for managed beans (also known as class-based beans).
+To support interception of producer methods and synthetic beans, the CDI specification includes an `InterceptionFactory`, which is a runtime oriented concept and therefore cannot be supported in Quarkus.
+
+Instead, Quarkus has its own API: `InterceptionProxy` and `@BindingsSource`.
```
I know that `@BindingsSource` is described below but it would not hurt to add a line in this paragraph too ...or remove `@BindingsSource` from this paragraph completely.

Ladicek replied:
Done.

mkouba commented on independent-projects/arc/runtime/src/main/java/io/quarkus/arc/InterceptionProxy.java:
```diff
@@ -0,0 +1,38 @@
+ * interceptors declared on the target class and its superclasses, are not supported.
+ * <p>
+ * The container provides a built-in bean with scope {@link jakarta.enterprise.context.Dependent Dependent}
+ * and qualifier {@link jakarta.enterprise.inject.Default Default} that can be injected.
```
```suggestion
 * and qualifier {@link jakarta.enterprise.inject.Default Default} that can be injected into method parameters of a producer method.
```

And we should probably also mention that synthetic beans can obtain a factory using the `SyntheticCreationalContext#getInterceptionProxy()` method.

Ladicek replied:
Done.

manovotn reviewed (approved):
LGTM now!

Ladicek commented:
I don't quite understand why the MP Telemetry TCK fails. @brunobat any ideas? Is that a known intermittent failure?

brunobat commented:
Should be intermittent. Looks like the exporter is trying to send data after the shutdown signal is sent, therefore hitting an executor that has been terminated already...
We had PRs to fix most these issues, but it seems it wasn't enough.

Ladicek commented:
OK, thanks! Merging this.

geoand commented:
> We had PRs to fix most these issues, but it seems it wasn't enough.

Interesting...
